django/db/models/constraints.py

```diff
@@ -643,6 +648,10 @@ def validate(self, model, instance, exclude=None, using=DEFAULT_DB_ALIAS):
                 # Ignore ordering.
                 if isinstance(expr, OrderBy):
                     expr = expr.expression
+                # Remove OpClass because it only has sense during the constraint
+                # creation.
+                if OpClass is not None and isinstance(expr, OpClass):
+                    expr = expr.get_source_expressions()[0]
```

Comment:
On reconsideration, `django.db.models` should not be aware of `django.contrib.postgres` or any other 3rd party packages and we already have 2 expression that should be omitted `OrderBy` and `OpClass`. Perhaps there will be more of them in the future, or maybe they even already exist in the ecosystem. Taking all of this into account, I'd recommend adding a new boolean property to `BaseExpression` instead, e.g. `constraint_validation_compatible`:
```python
class BaseExpression:
    ...
    # Can the expression be used during a constraint validation?
    constraint_validation_compatible = True
```

Reply:
We can raise a `ValueError` when more expressions are encapsulated in expressions with `constraint_validation_compatible` set to `False`.